web/src/RealmApp.js

```diff
@@ -0,0 +1,29 @@
+import React from "react";
+import * as Realm from "realm-web";
+
+const RealmAppContext = React.createContext();
+
+export const useRealmApp = () => {
+  const app = React.useContext(RealmAppContext);
+  if (!app) {
+    throw new Error(
+      `You must call useRealmApp() inside of a <RealmAppProvider />`
+    );
+  }
+  return app;
+};
+
+export const RealmAppProvider = ({ appId, children }) => {
+  const [app, setApp] = React.useState(new Realm.App(appId));
+  React.useEffect(() => {
+    setApp(new Realm.App(appId));
+  }, [appId]);
+  // Define count and refresh to hack around 
+  const [count, setCount] = React.useState(0);
+  function refresh() {
```

Comment:
One thing that might be less hacky than this refresh method is to set the user in state. This would accomplish the same functionality of forcing a rerender and give access to the logged-in user at any point. It would also make a lot more sense in the context of the app.  

For instance:

```
export const RealmAppProvider = ({ appId, children }) => {
  const [app, setApp] = React.useState(new Realm.App(appId));
  const [user, setUser] = React.useState(null);

  React.useEffect(() => {
    setApp(new Realm.App(appId));
  }, [appId]);
  return (
    <RealmAppContext.Provider value={{ app, setUser, user }}>
      {children}
    </RealmAppContext.Provider>
  );
};

```

and in LoginScreen.js (where you were formerly calling 'refresh()':
```

  const handleLogin = async () => {
    setError((e) => ({ ...e, password: undefined }));
    try {
      const isLoggedInUser = await app.logIn(
        Realm.Credentials.emailPassword(email, password)
      );
      setUser(isLoggedInUser);
    } catch (err) {
      handleAuthenticationError(err, setError);
    }
  };
```
and in Sidebar.js where you logout you could do:
```
        <UserDetails
          user={app.currentUser}
          handleLogout={async () => {
            await app.currentUser.logOut();
            setUser(null);
          }}
          handleEditPermissions={() => {
            setIsEditingPermissions(true);
          }}
        />
```

you could also make a wrapper around 'setUser' for added security, so children can't access parent state directly w/o some sort of validation.  

Let me know if this makes sense.

Reply:
Totally agreed that the `refresh()` method is hacky. I updated to use a similar approach to what you suggest, but I've encapsulated the user state inside of `RealmAppProvider` so that the API is just calling `logIn()` and `logOut()` with automatic state updates. Still a bit hacky since I needed to reconstruct the `Realm.App` object but it works!